Rewriting Tactic

TOTBWF opened the pull request:
## Patch Description

This PR adds a new tactic`rw`, which is a simple equality rewriting tactic, intended to be used in equational reasoning blocks.

### Motivation

When performing large equational reasoning proofs, it's extremely common to find yourself constructing sophisticated lambdas to pass into `cong`. This is extremely tedious, and can bog down large proofs in piles of verbose boilerplate. Normally one would use `rewrite` to avoid having to write out these complicated `cong` expressions, but this can obscure the proof, and is often impossible (IE: when using `Relation.Binary.Reasoning.Preorder`).

### Implementation

We solve this problem by introducing a new tactic, `rw`, that gives us a similar level succinctness as `rewrite`, but without it's restrictions. It operates by inferring the type of the goal, anti-unifying both ends of the equality, and then using that to construct the lambda argument to `cong`. This allows us to drastically reduce boilerplate; for instance
```
example : ∀ (m n : ℕ) → m ≡ n → suc (suc (m + 0)) + m ≡ suc (suc n) + (n + 0)
example m n eq = begin
    suc (suc (m + 0)) + m
  ≡⟨ cong (λ ϕ → suc (suc (ϕ + m))) (+-identityʳ m) ⟩
    suc (suc m) + m
  ≡⟨ cong (λ ϕ → suc (suc (ϕ + ϕ))) eq ⟩
    suc (suc n) + n
  ≡˘⟨ cong (λ ϕ → suc (suc (n + ϕ))) (+-identityʳ n) ⟩
    suc (suc n) + (n + 0)
  ∎
```

becomes

```
example : ∀ (m n : ℕ) → m ≡ n → suc (suc (m + 0)) + m ≡ suc (suc n) + (n + 0)
example m n eq = begin
    suc (suc (m + 0)) + m
  ≡⟨ rw (+-identityʳ m) ⟩
    suc (suc m) + m
  ≡⟨ rw eq ⟩
    suc (suc n) + n
  ≡˘⟨ rw (+-identityʳ n) ⟩
    suc (suc n) + (n + 0)
  ∎
```

This also plays nicely with `C-c C-m`, which allows for the quick generation of boilerplate for those who are hesitant of having tactics in the final source. 

### Notes

When the `rw` tactic is used in a hole, it doesn't report it's type properly when `C-c C-.` is used, and instead reports a metavariable. `C-c C-r` works fine though, so it's a minor quirk at worst.

MatthewDaggitt commented on src/Tactic/Rewrite.agda:
```diff
@@ -0,0 +1,222 @@
+    [] , ϕ
+
+  -- Helper for constructing applications of 'cong'
+  `cong : Term → Term → Term
```
So we should probably add this and `endpoints` to a new file `Relation.Binary.PropositionalEquality.Reflection`? That way they'd be reusable by other tactics in future?

I guess `endpoints` could take the error message string to throw as an extra argument?

MatthewDaggitt commented on src/Tactic/Rewrite.agda:
```diff
@@ -0,0 +1,222 @@
+
+  -- Helper for constructing applications of 'cong'
+  `cong : Term → Term → Term
+  `cong f eq = def (quote cong) (4 ⋯⟅∷⟆ vArg (lam visible (abs "ϕ" f)) ∷ 2 ⋯⟅∷⟆ vArg eq ∷ [])
```
You could use `vlam` instead of `lam visible` if you wanted?

MatthewDaggitt commented on src/Tactic/Rewrite.agda:
```diff
@@ -0,0 +1,222 @@
+
+private
+  -- Descend past a variable.
+  var-descend : ℕ → ℕ → ℕ
```
This is going to be a bit of a pain I know, but the standard library uses camel case (`varDescend`) for function names, rather than kebab case (`var-descend`). Any chance you could fix the names in this file?

MatthewDaggitt commented on src/Tactic/Rewrite.agda:
```diff
@@ -0,0 +1,222 @@
+      let (upats , ϕ') = patterns-descend ϕ pats in
+      (clause Γ upats (anti-unify ϕ' t t'))
+  anti-unify-clause ϕ (absurd-clause Γ pats) (absurd-clause Δ pats') =
+    Maybe.when (Γ =α=-Telescope Δ ∧ pats =α=-ArgsPattern pats') (absurd-clause Γ pats)
```
You should just be able to use `=α=` everywhere instead of `=α=-Telescope`, `=α=-ArgsPattern` etc.?

MatthewDaggitt commented on src/Tactic/Rewrite.agda:
```diff
@@ -0,0 +1,222 @@
+  ... | false = var ϕ []
+  anti-unify ϕ (sort unknown) (sort unknown) =
+    sort unknown
+  anti-unify ϕ (lit (nat n)) (lit (nat n')) with n Nat.≡ᵇ n'
```
Hmm can we collapse all these lit cases into the following?
```agda
anti-unify ϕ (lit l) (lit l') with l Literal.≡ᵇ l'
... | true  = lit l
... | false = var ϕ []
```

MatthewDaggitt reviewed (commented):
This looks awesome, thank you so much @TOTBWF! Added some comments.

I would be interested to see how it stacks up against #1158 which appears to be a bit dead in the water. In theory your solution is much nicer as it doesn't require you to manually point out where to apply `cong`. It would be great if you could try it out on the test cases in https://github.com/agda/agda-stdlib/pull/1158#issuecomment-613445321

So the big thing that is missing is a README file, under `README.Tactic.Rewrite` explaining how to use it. It also doubles as a test file, to prevent regressions. You could add the test cases there?

MatthewDaggitt commented:
Also I'm wondering slightly about the name, `rw` isn't terrible meaningful. In the past when we have a name that clashes with a builtin keyword, we usually surround it with angle brackets, e.g. `⟨rewrite⟩`. Welcome to alternative suggestions as well!

JacquesCarette reviewed (commented):
The level of commenting in this code is exemplary. I need to up my own game to match this!

gallais commented:
> Welcome to alternative suggestions as well!

Given that it's applying `cong`, I'd suggest a `cong`-based name. Perhaps `cong-` as we are already using `-`
to denote an argument that is not passed explicitly in e.g. [`-,_`](http://agda.github.io/agda-stdlib/Data.Product.html#2247)

That being said the missing argument in `-,_` is filled by unification (`-` is a textual representation of `_`) so
we could maybe use a different character to signal that it's a tactics. `cong!` maybe?

MatthewDaggitt commented on src/Reflection/Meta.agda:
```diff
@@ -16,7 +16,7 @@ import Relation.Binary.Construct.On as On
 
 open import Agda.Builtin.Reflection public
-  using (Meta) renaming (primMetaToNat to toℕ)
+  using (Meta) renaming (primMetaToNat to toℕ; primMetaEquality to _≡ᵇ_)
```
Err so the superscript `b` is for "boolean" so should stay. But to be honest, high-up on my list of priorities for v2.0 is to start introducing type classes into the library which should eliminate this clashing problem. For now qualifying it with `ℕ` is fine I think?

JacquesCarette replied:
It's obviously fine, I'm just sad to see the symmetry destroyed.

On the other point: I'm not sure introducing type classes is such a good idea. For `NonZero`, it made sense.

MatthewDaggitt replied:
> On the other point: I'm not sure introducing type classes is such a good idea. For NonZero, it made sense.

Interesting! Okay I look forward to your input about type classes when I kickstart the discussion. Hopefully will have time to do it around the new year.

MatthewDaggitt commented on src/Tactic/Rewrite.agda:
```diff
@@ -0,0 +1,222 @@
+  anti-unify-clause : ℕ → Clause → Clause → Maybe Clause
+  
+  anti-unify ϕ (var x args) (var y args') with x Nat.≡ᵇ y | anti-unify-args ϕ args args'
+  ... | _     | nothing    = var ϕ []
```
Eek, that's a bit far even for me :fearful:

JacquesCarette replied:
There's a reason I left it as a comment rather than a 'must change'.

The main point of (experimentally) changing the code to be like that would be to see the various patterns in the code appear more clearly. Right now, the code looks clear, but the patterns are not so easy to see.

To be clear: I'd recommend @TOTBWF experiment on his own branch with that, to see what emerges, if anything. But I'm not seriously suggesting changing it that way.

MatthewDaggitt replied:
Makes sense. Sorry, didn't mean to come across as judgy!

JacquesCarette replied:
Hey, your job title could be 'judger in chief' - in a good way! Your current role is to be final arbiter of the style of stdlib. You're doing a fine job of it, IMHO.

MatthewDaggitt commented:
I like `cong!`

TOTBWF commented:
Working through some of the test cases, I've stumbled across some really weird behaviour:
Consider this test case:
```agda
lit-test₁ : 40 + 2 ≡ 42
lit-test₁ = cong! refl
```

This produces unsolved metas: `refl` can't determine any of the implicits. This is odd though, as if we expand our the macro using `C-u C-c C-m`, there are no unsolved metas in sight.
```agda
lit-test₁ : 40 + 2 ≡ 42
lit-test₁ = cong (λ ϕ → ϕ) refl
```

It appears that Agda is checking the types of the arguments _before_ quoting them and passing them into the macro. This can be confirmed by adding a dummy argument to `refl`, like so:
```agda
lit-test₁ : 40 + 2 ≡ 42
lit-test₁ = cong! (refl {x = "uh-oh"})
```

Trying this out with the `quoteTerm` primitive, we can see the same behaviour: the term to be quoted is checked, _then_ quoted. To me, this seems pretty bizarre: macros will often provide enough context to make these unsolved metas disappear. Maybe there's some trick I'm missing that allows us to suppress these unsolved metas, but after a bit of experimentation I wasn't able to find one.

jespercockx commented:
> EDIT: Of course immediately after posting this I figured it out. The trick is to not quote the equality proof in question.

You mean you intend to change the type of `cong!` from `Term → Term → TC ⊤` to `x ≡ y → Term → TC ⊤`? Because that's what I would've suggested. Not quoting things unless it's really needed is a good principle to follow when implementing (Agda) macros.

TOTBWF commented:
So naive anti-unification is a bit too naive in a couple of cases. For instance, consider rewriting with `+-comm`: all subexpressions of the form `n + m` will get turned into `? + ?`. The original `cong!` tactic still seems useful because of its relative simplicity, but it probably makes sense to add a more sophisticated version that does some sort of subexpression search.

isovector commented:
This tactic is clearly very valuable as it stands, and I think it would be a shame to let perfection be the enemy of the good here. I'm exceptionally excited to start using it in my proofs.

MatthewDaggitt commented:
Agreed! @TOTBWF if you add the README and the CHANGELOG entries I'll merge it in and then afterwards we can work on improving it if people are still keen.

TOTBWF commented:
Cool! I've added the README file and fixed some minor issues. I'll stew a bit over what the best strategy for a more advanced version would be; I have some rough ideas, but nothing too concrete yet.

JacquesCarette commented:
Meant to comment on this earlier. Of course I agree with others that it is already valuable as it stands. What I would like to see is a bit of documentation about its limitations -- roughly so that people don't continually report known limitations and/or try to use it for what it is known not to work on.

I'm unsurprised that just anti-unification has limits. There's a reason things like AC and ACI-unification exist as separate items from just unification. You probably want to do a sort of anti-unification-modulo-theory. [No, this doesn't exist yet. I'm just giving it an obvious name.]

TOTBWF commented:
There's a short blurb in the README file that describes the situations it is likely to fail in, but ideally someone can come up with a smarter algorithm :)
